### PR TITLE
[Numpy] Numpy dstack

### PR DESCRIPTION
**Pull request by mikemwx**
## Description ##
Numpy compatible dstack with existing kernel from concat
[Numpy dstack](https://docs.scipy.org/doc/numpy/reference/generated/numpy.dstack.html)
### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [] Changes are complete (i.e. I finished coding on this PR)
- [] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes 
- Shape Inference added in src/operator/numpy/np_matrix_op.cc
- Forward and Backward compute in src/operator/nn/concat-in.h
- CPU and GPU registration in src/operator/numpy/np_matrix_op.cc and src/operator/numpy/np_matrix_op.cu
Pass CPU/GPU and lint tests.

#### Test Case Covered
- scalar tensor
- zero size tensor
- tensors with number of dimensions from 1 to 4

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here
Thank @haojin2 for guidance and reviewing!

### CODE REVIEW

File: src/operator/nn/concat-inl.h [hunk 1]
```diff
@@ -141,6 +141,37 @@ void ConcatCompute(const nnvm::NodeAttrs& attrs, const OpContext& ctx,
...
+  std::vector<TBlob> modified_inputs(inputs.size());
+  for (int i = 0; i < param.num_args; ++i) {
+    if (inputs[i].shape_.ndim() == 0) {
+      modified_inputs[i] = inputs[i].reshape(TShape(3, 1));
+    } else if (inputs[i].shape_.ndim() == 1) {
+      TShape t  = TShape(3, 1);
+      t[1] = inputs[i].shape_[0];
+      modified_inputs[i] = inputs[i].reshape(t);
+    } else if (inputs[i].shape_.ndim() == 2) {
+      TShape t  = TShape(3, 1);
```
Comment (haojin2) [inline src/operator/nn/concat-inl.h]: `TShape t = TShape(3, 1);`
one space between operators and operands, same for all applicable places.

File: src/operator/numpy/np_matrix_op.cc [hunk 2]
```diff
@@ -55,6 +55,67 @@ bool NumpyTransposeShape(const nnvm::NodeAttrs& attrs,
...
+    tmp[axis] = -1;
+    shape_assign(&dshape, tmp);
+  }
+
+  if (dshape.ndim() == -1) return false;
+  CHECK_NE(dshape.ndim(), 0) << "zero-dimensional arrays cannot be concatenated";
+
+  for (int i = 0; i < param_.num_args; ++i) {
+    CHECK(shape_assign(&(*in_shape)[i], dshape))
+        << "Incompatible input shape: expected " << dshape << ", got " << (*in_shape)[i];
```
Comment (haojin2) [inline src/operator/numpy/np_matrix_op.cc]:
```c++
    CHECK(shape_assign(&(*in_shape)[i], dshape))
      << "Incompatible input shape: expected " << dshape << ", got " << (*in_shape)[i];
```

Comment (haojin2) [reply src/operator/numpy/np_matrix_op.cc]: Same for all applicable places.

Comment (mikemwx) [reply src/operator/numpy/np_matrix_op.cc]: Thanks. Resolved.

**Comment by haojin2** [comment]
@mikemwx Please rebase with the latest numpy branch

**Comment by mikemwx** [comment]
renew a PR to master

**Pull request closed without merging**